cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	if err != nil {
+		return nil, err
+	}
+	if network == nil {
+		s.Info("network does not exist", "name", *s.GetServiceName(Network))
+		return nil, nil
+	}
+	return network.NetworkID, nil
+	//TODO: Support regular expression
+}
+
+// checkDHCPServerStatus checks the DHCP server status.
+func (s *PowerVSClusterScope) checkDHCPServerStatus() error {
+	dhcpID := *s.GetDHCPServerID()
+	dhcpServer, err := s.IBMPowerVSClient.GetDHCPServer(dhcpID)
+	if err != nil {
+		return err
+	}
+	if dhcpServer == nil {
+		return fmt.Errorf("error failed to get dchp server")
+	}
+	if *dhcpServer.Status != "ACTIVE" {
+		return fmt.Errorf("error dhcp server state is not active, current state %s", *dhcpServer.Status)
+	}
+	s.Info("DHCP server is found and its in active state")
+	return nil
+}
+
+// createDHCPServer creates the DHCP server.
+func (s *PowerVSClusterScope) createDHCPServer() (*string, error) {
+	dhcpServer, err := s.IBMPowerVSClient.CreateDHCPServer(&models.DHCPServerCreate{
+		DNSServer: pointer.String("1.1.1.1"),
+		Name:      s.GetServiceName(DHCPServer),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if dhcpServer == nil {
+		return nil, fmt.Errorf("created dhcp server is nil")
+	}
+	if dhcpServer.Network == nil {
+		return nil, fmt.Errorf("created dhcp server network is nil")
+	} else {
+		s.Info("DHCP Server network details", "details", *dhcpServer.Network)
+		s.SetStatus(Network, infrav1beta2.ResourceReference{ID: dhcpServer.Network.ID, ControllerCreated: pointer.Bool(true)})
+	}
+
+	return dhcpServer.ID, nil
+}
+
+// ReconcileVPC reconciles VPC.
+func (s *PowerVSClusterScope) ReconcileVPC() error {
+	// if VPC server id is set means the VPC is already created
+	vpcID := s.GetVPCID()
+	if vpcID != nil {
+		s.Info("VPC id is set", "id", vpcID)
+		vpcDetails, _, err := s.IBMVPCClient.GetVPC(&vpcv1.GetVPCOptions{
+			ID: vpcID,
+		})
+		if err != nil {
+			return err
+		}
+		if vpcDetails == nil {
+			return fmt.Errorf("error failed to get vpc with id %s", *vpcID)
+		}
+		s.Info("Found VPC with provided id")
+		//TODO(karthik-k-n): Set status here as well
+		return nil
+	}
+
+	// check vpc exist in cloud
+	vpcId, err := s.checkVPC()
+	if err != nil {
+		return err
+	}
+	if vpcId != "" {
+		s.SetStatus(VPC, infrav1beta2.ResourceReference{ID: &vpcId, ControllerCreated: pointer.Bool(false)})
+		return nil
+	}
+
+	// TODO(karthik-k-n): create a generic cluster scope/service and implement common vpc logics, which can be consumed by both vpc and powervs
+
+	// create VPC
+	s.Info("Creating a VPC")
+	vpcDetails, err := s.createVPC()
+	if err != nil {
+		return err
+	}
+	s.Info("Successfully create VPC")
+	s.SetStatus(VPC, infrav1beta2.ResourceReference{ID: vpcDetails, ControllerCreated: pointer.Bool(true)})
+	return nil
+}
+
+// checkVPC checks VPC exist in cloud.
+func (s *PowerVSClusterScope) checkVPC() (string, error) {
+	vpcDetails, err := s.getVPC()
+	if err != nil {
+		s.Error(err, "failed to get vpc")
+		return "", err
+	}
+	if vpcDetails == nil {
```

Comment:
Not required like dhcp get. I have checked, you can reverify once in vpc sdk get call.

Reply:
I think its required as vpc may not be found in cloud